src/db.c

```diff
@@ -160,13 +159,17 @@ robj *lookupKey(redisDb *db, robj *key, int flags, dictEntry **deref) {
             expire_flags |= EXPIRE_AVOID_DELETE_EXPIRED;
         if (flags & LOOKUP_ACCESS_EXPIRED)
             expire_flags |= EXPIRE_ALLOW_ACCESS_EXPIRED;
-        if (expireIfNeededWithSlot(db, key, expire_flags, key_slot) != KEY_VALID) {
+        if (expireIfNeeded(db, key, val, expire_flags) != KEY_VALID) {
             /* The key is no longer valid. */
             val = NULL;
+            if (link) *link = NULL;
         }
     }
 
     if (val) {
+        /* Shared objects can't be stored in the database. */
+        serverAssert(val->refcount != OBJ_SHARED_REFCOUNT);
```

Comment:
Does this mean that we might insert a shared robj? But it was found here?

Reply:
Yes. The new optimization that unifies key and value when they are small makes shared objects optimization less appealing.

Reply:
It is redundant. On insertion (dbAdd) we have explicit logic that unifies key & value into kvobj. I will remove it. Thanks!